<div style="text-align: right"> <b>Last Updated:</b> 8JUNE2020 </div>

# Clustering

__Authors:__ Dale Bowman, PhD; Natasha A Sahr, PhD

The goal of cluster analysis is to search for patterns in a data set by grouping the observations into clusters. In a good clustering, objects within clusters are similar and items in different clusters are dissimilar. For a data set with many features, we would like to use many relevant features to cluster the data. This makes clustering a type of <span><em>multivariate</em></span> analysis. Our variables are now a <span><strong>set</strong></span> of features, called <span><em>vectors</em></span>, that we will use to uncover latent structure in the data set. In cluster analysis we are looking for potential clusters (groups) that may (or may not) exist in the population.

## Distance Measures

Let $\mathbf{X}_i = (X_{1i}, X_{2i}, \ldots, X_{pi})^\prime$ be the $i$th observation vector. Whatever method of clustering we use, we must have some measure of similarity or dissimilarity in order to measure how close these multivariate measures are. One measure typically used is the simple Euclidean distance between two observations. The Euclidean distance between two $p$-dimensional observations, where $p$ is the number of features (variables) is 

$$d(\mathbf{X}_i,\mathbf{X}_j) = \sqrt{(\mathbf{X}_i-\mathbf{X}_j)^\prime}(\mathbf{X}_i-\mathbf{X}_j) = \sqrt\sum_{k=1}^{p} (X_{ki} - X_{kj}^2$$

## Hierarchical Clustering

Suppose we have $n$ observation vectors each consisting of $p$ different features. In hierarchical clustering we start with $n$ clusters (one for each observation). At each step one cluster is absorbed into another, ending up with only one cluster. This method is not based on a probability model but is a data analysis technique. It is an attempt to be computationally efficient instead of looking at all possible ways of partitioning $n$ items into $g$ groups which is approximatly $\frac{g^n}{g!}$. Even for moderate $g$ and $n$ this number is computationally prohibitive. Once an item enters a cluster it can not be removed from that cluster in the hierarchical method. This process can be done in reverse by starting with one cluster and splitting one cluster into two groups at each step until the final step results in $n$ groups.

_Agglomerative_ methods start with $n$ clusters and combine clusters at each step. The number of clusters shrinks and the clusters grow larger. _Divisive_ methods start with one cluster and at each step partition one cluster into two clusters, ending with $n$clusters. For each approach, a decision must be made as to the optimal number of clusters. Often the appropriate number of clusters can be found using dendrograms. Figure 1 below shows a dendrogram of an agglomerative hierarchical clustering of statistics on assault, murder, and rape in each of the 50 US states in 1973. It shows the order in which the clusters are combined based on the level of the bar connecting them. The lower the level the earlier the two clusters are combined. A horizontal line cutting through the dendrogram can be used to determine the number of clusters. For example, if a horizontal line were drawn at 6 on the vertical scale you would have 4 clusters.

In hierarchical clustering, at each step the distance between every two clusters is computed and the two closest clusters are combined. There are several options for defining the distance between clusters and each one can give slightly different clusters.

__Figure 1__

<body>
    <div class="img-box">
        <img src="images/den-1.jpg" alt="img1" style="width:100%" />
    </div>
</body>

### Example: Hierarchical Clustering

The data in the table below are based on crime rates per 100,000 population for certain US cities.

<table>
<thead>
<tr class="header">
<th align="left">City</th>
<th align="right">Murder</th>
<th align="right">Rape</th>
<th align="right">Robbery</th>
<th align="right">Assault</th>
<th align="right">Burglary</th>
<th align="right">Larceny</th>
<th align="right">Auto Theft</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Atlanta</td>
<td align="right">16.5</td>
<td align="right">24.8</td>
<td align="right">106</td>
<td align="right">147</td>
<td align="right">1112</td>
<td align="right">905</td>
<td align="right">494</td>
</tr>
<tr class="even">
<td align="left">Boston</td>
<td align="right">4.2</td>
<td align="right">13.3</td>
<td align="right">122</td>
<td align="right">90</td>
<td align="right">982</td>
<td align="right">669</td>
<td align="right">954</td>
</tr>
<tr class="odd">
<td align="left">Chicago</td>
<td align="right">11.6</td>
<td align="right">24.7</td>
<td align="right">340</td>
<td align="right">242</td>
<td align="right">808</td>
<td align="right">609</td>
<td align="right">645</td>
</tr>
<tr class="even">
<td align="left">Dallas</td>
<td align="right">18.1</td>
<td align="right">34.2</td>
<td align="right">184</td>
<td align="right">293</td>
<td align="right">1668</td>
<td align="right">901</td>
<td align="right">602</td>
</tr>
<tr class="odd">
<td align="left">Denver</td>
<td align="right">6.9</td>
<td align="right">41.5</td>
<td align="right">173</td>
<td align="right">191</td>
<td align="right">1534</td>
<td align="right">1368</td>
<td align="right">780</td>
</tr>
<tr class="even">
<td align="left">Detroit</td>
<td align="right">13.0</td>
<td align="right">35.7</td>
<td align="right">477</td>
<td align="right">220</td>
<td align="right">1566</td>
<td align="right">1183</td>
<td align="right">788</td>
</tr>
</tbody>
</table>

At each step in the clustering we will compute a distance matrix $\mathbf{D}$. For the first step, the distance between each pair of cities is computed using Euclidean distance to get the distance matrix below.

<table>
<thead>
<tr class="header">
<th align="left">City</th>
<th align="right">Atlanta</th>
<th align="right">Boston</th>
<th align="right">Chicago</th>
<th align="right">Dallas</th>
<th align="right">Denver</th>
<th align="right">Detroit</th>
</tr>
</thead>
<tbody>
<tr class="odd">
    <td align="left"><b>Atlanta</b></td>
<td align="right">0</td>
<td align="right">536.6</td>
<td align="right">516.4</td>
<td align="right">590.2</td>
<td align="right">693.6</td>
<td align="right">716.2</td>
</tr>
<tr class="even">
<td align="left"><b>Boston</b></td>
<td align="right">536.6</td>
<td align="right">0</td>
<td align="right">447.4</td>
<td align="right">833.1</td>
<td align="right">915.0</td>
<td align="right">881.1</td>
</tr>
<tr class="odd">
<td align="left"><b>Chicago</b></td>
<td align="right">516.4</td>
<td align="right">447.4</td>
<td align="right">0</td>
<td align="right">924.0</td>
<td align="right">1073.4</td>
<td align="right">971.5</td>
</tr>
<tr class="even">
<td align="left"><b>Dallas</b></td>
<td align="right">590.2</td>
<td align="right">833.1</td>
<td align="right">924.0</td>
<td align="right">0</td>
<td align="right">527.7</td>
<td align="right">464.5</td>
</tr>
<tr class="odd">
<td align="left"><b>Denver</b></td>
<td align="right">693.6</td>
<td align="right">915.0</td>
<td align="right">1073.4</td>
<td align="right">527.7</td>
<td align="right">0</td>
<td align="right"><font color="red"> 358.7 </font></td>
</tr>
<tr class="even">
<td align="left"><b>Detroit</b></td>
<td align="right">716.2</td>
<td align="right">881.1</td>
<td align="right">971.5</td>
<td align="right">464.5</td>
<td align="right"><font color="red"> 358.7 </font></td>
<td align="right">0</td>
</tr>
</tbody>
</table>

The smallest distance, <font color="red"> 358.7 </font>, is between Denver and Detroit. These two cities are joined to form $C_1 = \{\text{Denver},\text{Detroit}\}$. Next we will define a distance matrix between all cities and $C_1$. For this we need to define what we mean by the distance between clusters. Here we will use the _single linkage_ also known as nearest neighbor method, where the distance between two clusters $A$ and $B$ is defined as: 

$D(A,B) = \text{ minimum Euclidean distance between members in } A \text{ and members in } B$. 

Using this method we get the following distance matrix.

<table>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="right">Atlanta</th>
<th align="right">Boston</th>
<th align="right">Chicago</th>
<th align="right">Dallas</th>
<th align="right">$\mathbf{C_1}$</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><b>Atlanta</b></td>
<td align="right">0</td>
<td align="right">536.6</td>
<td align="right">516.4</td>
<td align="right">590.2</td>
<td align="right"><font color="blue"> 693.6 </font></td>
</tr>
<tr class="even">
<td align="left"><b>Boston</b></td>
<td align="right">536.3</td>
<td align="right">0</td>
<td align="right"><font color="red"> 447.4 </font></td>
<td align="right">833.1</td>
<td align="right"><font color="blue"> 881.1 </font></td>
</tr>
<tr class="odd">
<td align="left"><b>Chicago</b></td>
<td align="right">516.4</td>
<td align="right"><font color="red"> 447.4 </font></td>
<td align="right">0</td>
<td align="right">924.0</td>
<td align="right"><font color="blue"> 971.5 </font></td>
</tr>
<tr class="even">
    <td align="left"><b>Dallas</b></td>
<td align="right">590.2</td>
<td align="right">833.1</td>
<td align="right">924.0</td>
<td align="right">0</td>
<td align="right"><font color="blue"> 464.5 </font></td>
</tr>
<tr class="odd">
<td align="left">$\mathbf{C_1}$</td>
    <td align="right"><font color="blue"> 693.6 </font></td>
<td align="right"><font color="blue"> 881.1 </font></td>
<td align="right"><font color="blue"> 971.5 </font></td>
<td align="right"><font color="blue"> 464.5 </font></td>
<td align="right">0</td>
</tr>
</tbody>
</table>

The distance from Atlanta to $C_1$ is <font color="blue"> 693.6 </font> since the distance from Atlanta to Denver is 693.6 and from Atlanta to Detroit is bigger at 716.2 and the minimum distance is chosen. The distance from Boston to $C_1$ is similarly found: from Boston to Denver is 915.0 and from Boston to Detroit is 881.1. The numbers in blue in the distance matrix are the minimum from each city to Denver or Detroit. The smallest distance in the new matrix is between Boston and Chicago and so these are merged into $C_2 = \{\text{Boston}, \text{Chicago}\}$. 

The next distance matrix then becomes

<table>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="right">Atlanta</th>
<th align="right"><span class="math inline">$\mathbf{C_2}$</th>
<th align="right">Dallas</th>
<th align="right"><span class="math inline">$\mathbf{C_1}$</th>
</tr>
</thead>
<tbody>
<tr class="odd">
    <td align="left"><b>Atlanta</b></td>
<td align="right">0</td>
<td align="right">516.4</td>
<td align="right">590.2</td>
<td align="right">693.6</td>
</tr>
<tr class="even">
<td align="left"><span class="math inline">$\mathbf{C_2}$</td>
<td align="right">516.4</td>
<td align="right">0</td>
<td align="right">833.1</td>
<td align="right">881.1</td>
</tr>
<tr class="odd">
    <td align="left"><b>Dallas</b></td>
<td align="right">590.2</td>
<td align="right">833.1</td>
<td align="right">0</td>
<td align="right"><font color="red"> 464.5 </font></td>
</tr>
<tr class="even">
<td align="left"><span class="math inline">$\mathbf{C_1}$</td>
<td align="right">693.1</td>
<td align="right">881.1</td>
<td align="right"><font color="red"> 464.5 </font></td>
<td align="right">0</td>
</tr>
</tbody>
</table>

Now the smallest distance is <font color="red"> 464.5 </font> between Dallas and $C_1$, so the new cluster formed is $C_3 = \{\text{Dallas}, C_1\}$. The new distance matrix is now the one below.

<table>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="right">Atlanta</th>
<th align="right">$\textbf{C_2}$</th>
<th align="right">$\textbf{C_3}$</th>
</tr>
</thead>
<tbody>
<tr class="odd">
    <td align="left"><b>Atlanta</b></td>
<td align="right">0</td>
<td align="right"><font color="red"> 516.4 </font></td>
<td align="right">590.2</td>
</tr>
<tr class="even">
<td align="left">$\textbf{C_2}$</td>
<td align="right"><font color="red"> 516.4 </font></td>
<td align="right">0</td>
<td align="right">833.1</td>
</tr>
<tr class="odd">
<td align="left">$\textbf{C_3}$</td>
<td align="right">590.2</td>
<td align="right">833.1</td>
<td align="right">0</td>
</tr>
</tbody>
</table>

We combine Atlanta and $C_2$ since the distance between them is the smallest. The newly created cluster is $C_4 = \{\text{Atlanta},C_2\}$. 

The final distance matrix is then

<table>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="right">$\mathbf{C_4}$</th>
<th align="right">$\mathbf{C_3}$</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">$\mathbf{C_4}$</td>
<td align="right">0</td>
    <td align="right"><font color="red"> 590.2 </font></td>
</tr>
<tr class="even">
<td align="left">$\mathbf{C_3}$</td>
<td align="right"><font color="red"> 590.2 </font></td>
<td align="right">0</td>
</tr>
</tbody>
</table>

The final cluster is then $C_5 = \{C_3,C_4\}$. The dendrogram for this clustering is shown in Figure 2 below. The bar where two clusters are joined is at the level equal to the distance between them.

__Figure 2__

<body>
    <div class="img-box">
        <img src="images/denex-1.jpg" alt="img1" style="width:100%" />
    </div>
</body>

### Programming Example

To start to evaluate data, we need data. The `crime.csv` dataset contains 8 variables:

- `City`: US City
- `Murder`: number of murders per 100,000 people
- `Rape`: number of rapes per 100,000 people
- `Robbery`: number of robberies per 100,000 people
- `Assult`: number of assults per 100,000 people
- `Burglary`: number of burglaries per 100,000 people
- `Larceny`: number of larcenies per 100,000 people
- `Auto Theft`: number of auto thefts per 100,000 people

Import `pandas` as `pd`. Use the function `read_csv` in `pandas` to read in the `crime.csv` dataset and name it `dta_crime`. 

__Note__: The variable names are found in the file and the first column is the index column (`index_col=0`).

In [7]:
import pandas as pd

dta_crime = pd.read_csv("datasets/crime.csv", index_col=0)

Take a closer look at `dta_crime` by printing out data in a table.

In [8]:
print(dta_crime)

         Murder  Rape  Robbery  Assault  Burglary  Larceny  Auto Theft
City                                                                  
Atlanta    16.5  24.8      106      147      1112      905         494
Boston      4.2  13.3      122       90       982      669         954
Chicago    11.6  24.7      340      242       808      609         645
Dallas     18.1  34.2      184      293      1668      901         602
Denver      6.9  41.5      173      191      1534     1368         780
Detroit    13.0  35.7      477      220      1566     1183         788


To use the logic just learned, we could use the function `pdist`.  

In a `DataFrame` from `pandas` (named `pd`): include three arguments:

- the distance matrix: `squareform(pdist(dta_crime))`. The function `pdist` calculates the pairwise Euclidean distance. The function `squareform` puts these calculations in a matrix. 
- the column names: `columns = dta_crime.index`. 
- the row indices: `index = = dta_crime.index`. 

Print directly without saving. 

In [26]:
pd.DataFrame(
    squareform(pdist(dta_crime)),
    columns = dta_crime.index,
    index = dta_crime.index
)

City        Atlanta      Boston  ...       Denver     Detroit
City                             ...                         
Atlanta    0.000000  536.641910  ...   693.574113  716.196244
Boston   536.641910    0.000000  ...   914.978431  881.085807
Chicago  516.370042  447.403308  ...  1073.394769  971.527128
Dallas   590.175330  833.070837  ...   527.667253  464.467717
Denver   693.574113  914.978431  ...     0.000000  358.665373
Detroit  716.196244  881.085807  ...   358.665373    0.000000

[6 rows x 6 columns]

If we decide to use this route, we still need to iteratively update the distance matrix using single linkage. This seems like a lot of work! Instead, use `linkage` from `scipy.cluster.hierarchy` to do the heavy lifting.

To start, from `scipy.cluster.hierarchy` import `linkage`. 

In [ ]:
from scipy.cluster.hierarchy import linkage

Use the function `linkage` where the first argument is our data `dta_crime` and the second argument is the type of linkage `single`. Name this object `Z`. 

In [30]:
Z = linkage(dta_crime, 'single')

Using the `print` function, print the contents of `Z`.

In [61]:
print(Z)

[[  4.           5.         358.66537329   2.        ]
 [  1.           2.         447.40330799   2.        ]
 [  3.           6.         464.46771685   3.        ]
 [  0.           7.         516.37004173   3.        ]
 [  8.           9.         590.17532988   6.        ]]


Each row of the object `Z` is an interation in our single linkage method. The first row corresponds to the first distance matrix. It identifies the smallest distance, `358.665`, in the third colum and states that it results in a merger of observation 4 (`Denver`) and 5 (`Detroit`). _Remeber that the indices start at 0._ The fourth column shows us that this results in 2 clusters/nodes. 

While this is an improve approach, we really want the dendrogram. To get the dendrogram in without pre-calculations, we import `plotly.figure_factory`.

To start, import `plotly.figure_factory` as `ff`. 

In [ ]:
import plotly.figure_factory as ff

Use the `create_dendrogram` function in `plotly.figure_factory` with three arguments:

- `dta_crime`: the data
- `labels=dta_crime.index`: the labels of the observations
- `linkagefun=lambda x: linkage(dta_crime, 'single', metric='euclidean')`: the type of linkage, single linkage with Euclidean distance

Name this object `fig_crime`. 

In [54]:
fig_crime = ff.create_dendrogram(dta_crime, 
                           labels=dta_crime.index,
                           linkagefun=lambda x: linkage(dta_crime, 'single', metric='euclidean'))

Update the `width` and `height` for the plot `fig_crime` using the function `update_layout`. Set the `width=600` and `height=400`. This allows us to see the dendrogram better.

In [ ]:
fig_crime.update_layout(width=600, height=400)

Use the function `show` to display the plot. Does it match the worked example?

In [ ]:
fig_crime.show()

This was a unsupervised clustering method. Welcome to the world of Machine Learning!

## Nonhierarchical Methods

### Partitioning

_Partitioning_ methods are also called _optimization methods_. In this approach to clustering observations are separated into $g$ clusters and optimal clustering is chosen according to some criteria. One partitioning method commonly used is _$k$-means_. 

In $k$-means clustering, an initial $g$ observations are chosen as seeds. Here $g$ is the number of clusters we want to use. The seeds are typically chosen at random so that they are a specified distance apart. The remaining observations are assigned to the cluster whose seed they are closest to. When all points (observation vectors) have been assigned, each point is re-examined to see if it is closer to the _centroid_ of any other cluster. The centroid is the average, over all the observations in the cluster, of each of the $p$ features. If points are closer to a centroid from a different cluster, they are moved to that cluster. Centroids are re-calculated and the process continues until there is no further improvement. If there is no convergence or slow convergence there may be no natural clustering in the data. $k$-means requires that the number of clusters be specified in advance but it allows observations to move between clusters if a better clustering is found. Like hierarchical clustering, $k$-means is not based on probability measures.

We have previously used the `iris.csv` dataset. The `iris.csv` dataset contains 5 variables:

- `SepalLength`: the sepal length (cm)
- `SepalWidth`: the sepal width (cm)
- `PetalLength`: the petal length (cm)
- `PetalWidth`: the petal width (cm)
- `Species`: the flower species (cm)

Import `pandas` as `pd`. Use the function `read_csv` in `pandas` to read in the `iris.csv` dataset and name it `dta_iris`. 

__Note__: The variable names are not found in the file and the file does not have an index.

Print the first 5 observations to familiarize yourself with the data again.

In [62]:
dta_iris = pd.read_csv("datasets/iris.csv", 
                       header=None,
                       names=['SepalLength',
                              'SepalWidth',
                              'PetalLength',
                              'PetalWidth',
                              'Species'])

dta_iris.head(5)

SepalLength  SepalWidth  PetalLength  PetalWidth      Species
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

We are going to use `SepalLength`, `SepalWidth`, `PetalLength`, and `PetalWidth` to perform $k$-means. Subset the dataset `dta_iris` using the `iloc` function and call the new dataset `dta_iris_new`.

In [63]:
dta_iris_new = dta_iris.iloc[:, [0,1,2,3]]

To perform $k$-means, from `sklearn.cluster` import `KMeans`. We will need this function. 

In [64]:
from sklearn.cluster import KMeans

We will arbitarily assign the number of clusters, $k$, as 5. Implementent $k$-means clustering using `KMeans` function and the argument `n_clusters=5`. Name this object `kmeans_iris_5`.

In [65]:
kmeans_iris_5 = KMeans(n_clusters=5)

Now, we need to connect the $5$-means method to our data, `dta_iris_new`. Use the function `fit_predict` on the object `kmeans_iris_5` using the argument `dta_iris_new`. Name this object `y_kmeans_iris_5`. Print `y_kmeans_iris_5` to view its contents.

In [66]:
y_kmeans_iris_5 = kmeans_iris_5.fit_predict(dta_iris_new)
print(y_kmeans_iris_5)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 3 3 3 0 3 3 3 0 3 0 0 3 0 3 0 3 3 0 3 0 3 0 3 3
 3 3 3 3 3 0 0 0 0 3 0 3 3 3 0 0 0 3 0 0 0 0 0 3 0 0 2 3 4 2 2 4 0 4 2 4 2
 2 2 3 2 2 2 4 4 3 2 3 4 3 2 4 3 3 2 4 4 4 2 3 3 4 2 2 3 2 2 2 3 2 2 2 3 2
 2 3]


These are the cluster assignments to each flower, corresponding to each row of the dataset. At a column to your dataset `dta_iris_new` called `cluster`. containing the cluster assignments. 

In [77]:
dta_iris_new['cluster'] = y_kmeans_iris_5

Use the function, `astype('category')` to change the variable `cluster` into a categorical variable representing cluster assignment.

In [78]:
dta_iris_new['cluster'] = dta_iris_new['cluster'].astype('category')

Import `plotly.express` as `px` to use the `scatter` function.

In [79]:
import plotly.express as px

If we want to evaluate the relationship of sepal width on sepal length, we will have the following arguments:

- `dta_iris_new`
- `x="SepalWidth"`
- `y="SepalLength"`

We can use the additional argument `color="cluster"` to see the affect of the clustering. 

Make sure the plot displays nicely by using arguments    
- `template = "simple_white"`
- `opacity = 0.4`
- `trendline="ols"`

Use the function `scatter` in `plotly.express` and the three arguments to create a scatterplot figure names `fig_iris_cluster`. 

Using the function `update_layout` have the title display `"Relationship between Sepal Length and Sepal Width"` and show the plot `fig_iris_cluster` using the function `show`. 

In [80]:
fig_iris_cluster = px.scatter(dta_iris_new, 
                              x="SepalWidth", 
                              y="SepalLength",
                              color="cluster",
                              template = "simple_white",
                              opacity = 0.4,
                              trendline="ols")

fig_iris_cluster.update_layout(title_text="Relationship between Sepal Length and Sepal Width")

fig_iris_cluster.show()

Do you see a reason for the clusters?

### Choosing the number of clusters

Oftentimes there are suggestions from the domain of the problem that suggest an appropriate number of clusters, $g$.  In many instances however, we must make our own decisions about the __best__ value of $g$.  One method is to select a set of cluster sizes that may be appropriate and compare the clustering using sum of squares to see which value of $g$ gives the best clustering.  Ideally we would like the distances between observation vectors to be small inside each cluster and large across clusters.  We can evaluate these using within sum of squares matrices and between sum of squares matrices.  In an ideal clustering, the within SS matrix should be __small__ and the between SS matrix should be __large__.  There are various measures we can use to assess the relative size of matrices, including the trace and determinant.


To evaluate the number of clusters, we can use the Elbow method. Copy and paste the code below. 

    
<blockquote><tt>
<pre>Error = []
max_cluster = 11
for i in range(1, max_cluster):
    kmeans = KMeans(n_clusters = i).fit(dta_iris_new)
    kmeans.fit(dta_iris_new)
    Error.append(kmeans.inertia_)
    
find_cluster = pd.DataFrame({'Error': Error, 
                             'Cluster': list(range(1,max_cluster))})

fig_iris_cluster = px.line(find_cluster,
                     x="Cluster", 
                     y="Error", 
                     template = "simple_white")

fig_iris_cluster.show()</pre>
</tt></blockquote>

In [121]:
Error = []
max_cluster = 11
for i in range(1, max_cluster):
    kmeans = KMeans(n_clusters = i).fit(dta_iris_new)
    kmeans.fit(dta_iris_new)
    Error.append(kmeans.inertia_)
    
find_cluster = pd.DataFrame({'Error': Error, 
                             'Cluster': list(range(1,max_cluster))})

fig_iris_cluster = px.line(find_cluster,
                     x="Cluster", 
                     y="Error", 
                     template = "simple_white")

fig_iris_cluster.show()

The point that forms the elbow of the plot above is approximately $k=3$. Suggesting the optimum number of clusters is 3.  

Repeat the procedure to generate a dataset from $3$-means procedure. Use the following steps:

1. Use the `KMean` function to set `n_clusters=3` and name the object `kmeans_iris_3`.
2. Call the `fit_predict` function on `kmeans_iris_3` using `dta_iris_new` and name the object `y_kmeans_iris_3`. 
3. Add a column `cluster-3` to the `dta_iris_new` dataset using the variable `y_kmeans_iris_3`.
4. Update the variable type of `cluster-3` using the function `astype('category')`. 

In [126]:
kmeans_iris_3 = KMeans(n_clusters=3)
y_kmeans_iris_3 = kmeans_iris_3.fit_predict(dta_iris_new)
dta_iris_new['cluster-3'] = y_kmeans_iris_3
dta_iris_new['cluster-3'] = dta_iris_new['cluster-3'].astype('category')

Update the plot by coloring by `cluster-3`. Name the new resulting plot `fig_iris_cluster_3`. Update and show the plot `fig_iris_cluster_3` as before. 

In [131]:
fig_iris_cluster_3 = px.scatter(dta_iris_new, 
                              x="SepalWidth", 
                              y="SepalLength",
                              color="cluster-3",
                              template = "simple_white",
                              opacity = 0.4,
                              trendline="ols")

fig_iris_cluster_3.update_layout(title_text="Relationship between Sepal Length and Sepal Width")

fig_iris_cluster_3.show()

### Mixtures of Distributions

We can use probability models to perform clustering if we assume that there are $g$ multivariate normal distributions corresponding to the proposed $g$ clusters.  We want to assign each observation vector to the distribution that is most probable.  We assume that each observation vector $\mathbf{X}$ has a probability distribution given by the mixture:
$$h(\mathbf{X}) = \sum_{i=1}^g \alpha_i f_i(\mathbf{X}, \mathbf{\mu}_i, \mathbf{\Sigma}_i)$$
where $0 \le \alpha_i \le 1$ and $\displaystyle \sum_{i=1}^g \alpha_i = 1$.  $f_i(\cdot)$ is a probability distribution usually assumed to be a multivariate normal distribution with mean vector $\mathbf{\mu}_i$ and covariance matrix $\mathbf{\Sigma}_i$.  We assign an observed vector to the cluster with highest posterior probability.